### DISCUSSION TITLE
Can we use sn.mean() without providing the bounds?

### DISCUSSION THREAD

**Discussion by aitsam12** [opendp/opendp#161] [SmartNoise] [answered]
In sn.analysis, is it possible to conduct the analysis without providing bounds while calling the query and while pre-processing the data (clamp, resize, impute)?
If its possible, can someone tell me how bounds are calculated automatically?
If its not possible, can someone tell me where it is mandatory to provide bounds and why?

**Comment by Shoeboxam** [comment, marked as answer]
There is no feature for automatic bounds calculation in SmartNoise. However, if you clamp, the bounds will be propagated up through the computation. 

The reason why bounds are required is because the noise scale is calibrated based on the maximum influence that modifying a user can have on the non-private statistic. For instance, if you are computing the sum, adding or removing one user can influence the sum by at most `max(|lower_bound|, |upper_bound|)`. In other words, what is the maximum the query can change, should we change one user? This is called the "sensitivity" of the query. If the data is unbounded, then the sensitivity is also unbounded, and the necessary noise scale to make the query differentially private is undefined. Ultimately, any query whose sensitivity depends on data bounds needs bounded data in order to make a differentially private release of the result.

There are other libraries that do automatic bounds estimation. That would be a good feature request for the OpenDP library.

**Reply by gonzalo-munillag** [reply]
That is a great explanation @Shoeboxam. Then, anyone who uses Smartnoise will violate the privacy of the people behind the dataset because they need to peak into the max and min values to make the query. Is there a clever way to avoid this, even when one does not  know beforehand that the attribute could be e.g. age (0-110)? (Also not considering the runtime is a blackbox)

**Reply by Shoeboxam** [reply]
There's only privacy violation if you look at the data to get the bounds. In many situations a reasonable guess is good enough- if I'm requesting the mean age, it's pretty reasonable to plug in bounds of [0, 100] without ever seeing the data itself. Another example sensitivity- the known-n addition/removal restricted-sensitivity for the mean is `(upper - lower) / n / 2`. You can get a rough feel from these example sensitivities how the accuracy drops as the bounds get wider. In general I err towards wider bounds, to get unbiased releases. If you are curious, there are more sensitivities in the smartnoise-core repository in the whitepapers folder.

It would probably be good for me to explain the bias-variance tradeoff of preprocessing as well. For clamping, if your bounds are too wide, you'll get an unbiased estimate with greater variance. If your bounds are too tight, you'll get a biased answer with less variance. For resize, you'll get an answer biased toward the imputation parameters if you provide an n that is too large, and greater variance if you underestimate n. Relatedly, the accuracy estimates on the library only account for the variance introduced by the mechanism. Any increase in bias or variance introduced by preprocessing (clamp, impute, resize) cannot be captured by the accuracy estimate as this would make the accuracy estimate itself sensitive.

**Reply by Shoeboxam** [reply]
Of course you can get a DP estimate of N with much greater utility than a DP estimate of the min and max. It is possible to use the quantile to release DP estimates of min and max. Unfortunately, you need to provide lower and upper bounds for what you think the min is, and lower and upper bounds for what you think the max is. So that approach is turtles all the way down!
